src/server/fts/storage.rs

```diff
@@ -0,0 +1,209 @@
+    PathBuf::from(path.as_bytes())
+}
+
+async fn file_read<T>(
+    trussed: &mut T,
+    platform: &[u8],
+    filename: &str,
+) -> Result<Vec<u8>, CoapError>
+where
+    T: trussed::client::FilesystemClient,
+{
+    trussed::try_syscall!(trussed.locate_file(
+        Location::Internal,
+        None,
+        get_filename(platform, filename)
+    ))
+    .map_err(|_| {
+        error!("Storage error when trying to locate {}", filename);
+        CoapError::internal("Internal error")
+    })?
+    .path
+    .ok_or_else(CoapError::not_found)?;
+
+    let path = get_path(platform, filename);
+
+    Ok(
+        trussed::try_syscall!(trussed.read_file(Location::Internal, path))
+            .map_err(|_| {
+                error!("Storage error when trying to read {}", filename);
+                CoapError::internal("Internal error")
+            })?
+            .data
+            .to_vec(),
+    )
+}
+
+async fn file_write<T>(
+    trussed: &mut T,
+    platform: &[u8],
+    filename: &str,
+    data: &[u8],
+) -> Result<bool, CoapError>
+where
+    T: trussed::client::FilesystemClient,
+{
+    let path = get_path(platform, filename);
+    let exists = trussed::try_syscall!(trussed.locate_file(
+        Location::Internal,
+        None,
+        get_filename(platform, filename)
+    ))
+    .map_err(|_| {
+        error!("Storage error when trying to locate {}", filename);
+        CoapError::internal("Internal error")
+    })?
+    .path
+    .is_some();
+
+    trussed::try_syscall!(trussed.write_file(
+        Location::Internal,
+        path,
+        Message::from_slice(data).map_err(|_| CoapError::internal("File is too big"))?,
+        None
+    ))
+    .map_err(|_| {
+        error!("Storage error when trying to write {}", filename);
+        CoapError::internal("Internal error")
+    })?;
+
+    Ok(exists)
+}
+
+async fn file_delete<T>(trussed: &mut T, platform: &[u8], filename: &str) -> Result<bool, CoapError>
+where
+    T: trussed::client::FilesystemClient,
+{
+    let path = get_path(platform, filename);
+    let exists = trussed::try_syscall!(trussed.locate_file(
+        Location::Internal,
+        None,
+        get_filename(platform, filename)
+    ))
+    .map_err(|_| {
+        error!("Storage error when trying to locate {}", filename);
+        CoapError::internal("Internal error")
+    })?
+    .path
+    .is_some();
+
+    if !exists {
+        return Ok(false);
+    }
+
+    trussed::try_syscall!(trussed.remove_file(Location::Internal, path)).map_err(|_| {
+        error!("Storage error when trying to delete {}", filename);
+        CoapError::internal("Internal error")
+    })?;
+
+    Ok(true)
+}
+
+pub async fn handler(
+    request: Request<Endpoint>,
+    state: &ServerState,
+    client: Arc<Mutex<CriticalSectionRawMutex, Client>>,
+) -> Result<Response, CoapError> {
+    let client = client.lock().await;
+    if request.unmatched_path.len() != 1 {
```

Comment:
This catches all illegal file names (`.`, `..`, something with `/`) and returns 4.04 even for PUT and DELETE.

Reply:
Actually, the client did some strange transformation on path and didn't sent neither `.` nor `..`, and `/` in my testing was treated as path separator and not a character in file name. When those characters are percent-encoded, it returns expected error codes.